### PR TITLE
refactor: allow different tables Libdbmx TrieDB implementation using generics

### LINKED ISSUE #359: `Trie` should receive a `Db` trait instead of a path with the db location
See this comment: https://github.com/lambdaclass/ethereum_rust/pull/316/files#r1750443225

Instead we could pass a generic DB trait that has all the functions needed to query and update a db, like `get`, `insert`, `remove`.

For example, we could have a `StorageTrieDb` implementation of `DB` which stores the trie in the `StorageDb` table. Or we could have an `InMemory` implementation which stores the trie in memory.

### PR DESCRIPTION
**Pull request by Arkenan**
**Motivation**

This PR allows us to have multiple tries, all using the same database, but different tables, so they can't collide with each other even if they were to share key space.

**Changes**

- Makes the `Libmdbx` implementation for the `TrieDB` trait generic. The type is a Table struct which specifies which table will be used in the trie.
- The `Libmdbx` struct now has a borrowed db instead of a moved one. This allows us to use a single db struct in several tries, which didn't work in the previous model (`resource temporarily unavailable`).
- Adds `test_utils` for creating a db. Most of the PR code is just changing tests, but the Trie code itself is 100% unchanged.

Closes #359

### CODE REVIEW

File: crates/storage/trie/db/libmdbx.rs [hunk 1]
```diff
@@ -1,70 +1,90 @@
...
 use crate::error::StoreError;
 use libmdbx::{
-    orm::{table, Database},
+    orm::{table, Database, Table},
     table_info,
 };
 
 /// Libmdbx implementation for the TrieDB trait, with get and put operations.
-pub struct Libmdbx(Database);
+pub struct Libmdbx<'a, T: Table> {
```
Comment (fmoletta) [inline crates/storage/trie/db/libmdbx.rs:10]: We will be importing this type outside the trie module, so I think we should use a less generic name for it, maybe something like `LibmdbxTrieDB`

Comment (Arkenan) [reply crates/storage/trie/db/libmdbx.rs]: Good idea! Done.

File: crates/storage/trie/db/libmdbx.rs [hunk 2]
```diff
@@ -1,70 +1,90 @@
...
-    }
-
-    #[cfg(test)]
-    /// Creates a temporary DB, for testing purposes only
-    pub fn init_temp() -> Self {
-        use tempdir::TempDir;
-        let tables = [table_info!(Nodes)].into_iter().collect();
-        Self(Database::create(None, &tables).expect("Failed to create temp DB"))
+impl<'a, T: Table> LibmdbxTrieDB<'a, T> {
+    pub fn new(db: &'a Database) -> Self {
```
Comment (mpaulucci) [inline crates/storage/trie/db/libmdbx.rs:18]: have you tried using an Rc instead of lifetimes?

Comment (fmoletta) [reply crates/storage/trie/db/libmdbx.rs:18]: Adding to this: In my PR I had to do some refactors to make the trait work with our StoreEngine api, and I replaced the lifetime with an `Arc`. #399

Comment (fmoletta) [reply crates/storage/trie/db/libmdbx.rs:18]: Using an `Rc` would force us to use an `Rc` in the libmdbx `Store`, making it not `Send`

Comment (mpaulucci) [reply crates/storage/trie/db/libmdbx.rs:18]: @fmoletta so we should merge this one as-is since you make the changes in your PR?

Comment (fmoletta) [reply crates/storage/trie/db/libmdbx.rs:18]: Sure

**Pull request merged**